### PR TITLE
Nilable classes in spec, part 2

### PR DESCRIPTION
**Pull request by mppf**
Follow-on to PR #14002.

* Fill in the Nilable Classes section of the spec
* Add some details to an Implicit Class Conversion section
* Fix up param conversions section

Reviewed by @vasslitvinov - thanks!

### CODE REVIEW

File: spec/Classes.tex [hunk 1]
```diff
@@ -307,12 +370,12 @@ \subsection{Class Values}
 \end{chapelcompopts}
```
Comment (vasslitvinov) [inline spec/Classes.tex]: Could you change Line 358 above to say:
`... is deleted automatically when it is not`

File: spec/Classes.tex [hunk 2]
```diff
@@ -271,6 +267,72 @@ \subsection{Class Types}
...
+\end{chapel}
+\begin{chapeloutput}
+\end{chapeloutput}
+\end{chapelexample}
+
+\label{Methods_On_Nilable_In_Prototype_Modules}
+
+Within a \chpl{prototype} module, the compiler will implicitly convert
+a nilable method receiver to the non-nilable type by adding a \chpl{!}
+call.
```
Comment (vasslitvinov) [inline spec/Classes.tex]: I think the `borrow` method is an exception. It is applicable to `nil` and if so produces `nil`.

If this is the case, please explain it in this subsection.

File: spec/Classes.tex [hunk 3]
```diff
@@ -307,12 +370,12 @@ \subsection{Class Values}
...
-class \chpl{C}.  The declaration of variable \chpl{c2} shows that these steps can
-be combined.  The type of \chpl{c2} is also \chpl{C}, determined implicitly from
-the the initialization expression.  Finally, an object of type \chpl{D} is created and
-assigned to \chpl{c}.
+When the variable \chpl{c} is declared, it initially has the value of
+\chpl{nil}.  The next statement assigned to it an instance of the class
+\chpl{C}.  The declaration of variable \chpl{c2} shows that these steps
+can be combined.  The type of \chpl{c2} is also \chpl{C}, determined
+implicitly from the the initialization expression.  Finally, an object
+of type \chpl{D} is created and assigned to \chpl{c}.
```
Comment (vasslitvinov) [inline spec/Classes.tex]: This example lacks clarity about nilability of `c2`. Depending on the behavior of `.borrow()` - see my earlier comment - need to add or edit that:
* the type of `c2` is nilable `is also \chpl{C?}`, or
* the above code would produce an error upon `c.borrow()` if it were in a non-prototype module.

Oh, either way, the type of `c2` is actually `borrowed`, so need to remove the `also`.

File: spec/Classes.tex [hunk 4]
```diff
@@ -359,7 +422,79 @@ \subsection{Class Methods}
 \index{secondary methods}
 
 Methods on classes are referred to as to as \emph{class methods}.
-See the methods section~\rsec{Methods}.
+See the methods section~\rsec{Methods} for more information about methods.
+
+Within a class method, the type of \chpl{this} is generally the
+non-nilable \chpl{borrowed} variant of the class type. It is different
+for type methods (see below) and it might be a different type if the
+class method is declared with as a secondary method with a type
```
Comment (vasslitvinov) [inline spec/Classes.tex]: remove `with` after `is declared`

File: spec/Classes.tex [hunk 5]
```diff
@@ -359,7 +422,79 @@ \subsection{Class Methods}
 \index{secondary methods}
 
 Methods on classes are referred to as to as \emph{class methods}.
-See the methods section~\rsec{Methods}.
+See the methods section~\rsec{Methods} for more information about methods.
```
Comment (vasslitvinov) [inline spec/Classes.tex]: Or simply:

`Methods on classes (\rsec{Methods}) are referred to as to as \emph{class methods}.`

?

File: spec/Classes.tex [hunk 6]
```diff
@@ -271,6 +267,72 @@ \subsection{Class Types}
...
+nilable. To declare a nilable class type, use the \chpl{?} operator to
+create a nilable type. For example, if \chpl{C} is a class type, then
+\chpl{C?} indicates nilable the class type with generic management. The
+\chpl{?} operator can be combined with memory management specifiers as
+well. For example, \chpl{borrowed C?} indicates a nilable class using the
+\chpl{borrowed} memory management strategy. Note that the \chpl{?}
+operator only applies to types.
+
+The postfix \chpl{!} operator applies to a type or a value. When applied
+to a type, it returns the non-nilable borrowed type. When applied to a
```
[the remark this answers is not in the document]
Comment (mppf) [reply spec/Classes.tex]: Oh, yeah, that's true. I think I can adjust the spec for it.

File: spec/Conversions.tex [hunk 7]
```diff
@@ -217,9 +266,9 @@ \subsection{Explicit Numeric Conversions}
 \index{conversions!explicit!numeric}
 
 Explicit conversions are allowed from any numeric type, boolean, or
-string to any other numeric type, boolean, or string.  
+string to any other numeric type, boolean, or string.
 
-% A valid \chpl{bool} value behaves like a single unsigned bit.  
+% A valid \chpl{bool} value behaves like a single unsigned bit.
 When a \chpl{bool} is converted to a \chpl{bool}, \chpl{int}
```
Comment (vasslitvinov) [inline spec/Conversions.tex]: The two changes above may conflict with Engin's #13996.
Just sayin'.

File: spec/Expressions.tex [hunk 8]
```diff
@@ -298,6 +295,14 @@ \section{Precedence and Associativity}
 \hline
 \verb@new@ & right & initializer call \\
 \hline
+\verb@owned@ & right & apply management strategy to a class \\
+\verb@shared@ & right & \\
+\verb@borrowed@ & right & \\
+\verb@unmanaged@ & right & \\
```
Comment (vasslitvinov) [inline spec/Expressions.tex]: Let's leave just one `right` -- at `shared` or `owned` -- and remove it for the other management strategies.

File: spec/Expressions.tex [hunk 9]
```diff
@@ -298,6 +295,14 @@ \section{Precedence and Associativity}
...
+\hline
+\verb@?@ & left & compute a nilable class type \\
+postfix \verb@!@ & left & assert non-nilable and borrow \\
```
Comment (vasslitvinov) [inline spec/Expressions.tex]: prepend `postfix` to `?` as well, for consistency
remove `left` in the second line above

File: spec/Methods.tex [hunk 10]
```diff
@@ -163,6 +162,12 @@ \section{The Method Receiver and the {\em this} Argument}
 methods}.  Methods may also be defined to have \chpl{type}
 receivers---these are known as \emph{type methods}.
 
+Note that within a method for a class \chpl{C}, the type of \chpl{this}
+is generally \chpl{borrowed C}. Within a type method on a class \chpl{C},
+\chpl{this} refers to a class type with management and nilability matching
```
Comment (vasslitvinov) [inline spec/Methods.tex]: add `C` : `refers to the class type \chpl{C}`

File: spec/Classes.tex (hunk 5 above)
Comment (mppf) [reply spec/Classes.tex]: I think I like the longer one better on this one.

File: spec/Methods.tex (hunk 10 above)
Comment (mppf) [reply spec/Methods.tex]: I'm not following?

Comment (mppf) [reply spec/Methods.tex]: Oh, I got it.

**Pull request merged**